### ISSUE TITLE
TiddlyHost Won't Save

### ISSUE THREAD

**Issue by LordPalandus** [simonbaird/tiddlyhost#345]
The issue first started up last week. I hit the red circle to save, and it won't save. No matter how long I wait. Sometimes it gives an error other times no error at all, for hours. Literally, not figuratively. 

Before that, it would save in about 5-10 seconds. Sometimes faster, rarely more than 15 seconds. 

As such, I've not been able to work on my sites, for fear I might lose data when I hit save and have a power outage. 

Pages open up fast enough and can open up tiddlers within a tiddlywiki without issue, but saving changes, takes forever. 

The times it does give an error, it says there was a timeout. Not sure why that would be as I have a 300 megabit down, and 75 megabit up connection, and the only time I've remembered getting timeouts was back in the olden days of 1 megabit down connections DSL.

**Comment by sobjornstad** [comment]
I've been trying to set up some stuff in TiddlyHost this week and it seems to me like the servers are under very heavy load right now – quite a few things have just taken way longer than they ought to (e.g., the “verify your email” message when I registered a new account took >30 minutes to arrive). Maybe worth looking into.

**Comment by simonbaird** [comment]
Thanks for the report. I'm not sure what the cause is, but I will look into it. Can you share what version of TiddlyWiki and what browser you're using?

**Comment by LordPalandus** [comment]
I'm using Windows 8.1 and Google Chrome. Didn't have issues with save/load a few weeks ago, but can't reliably save/load right now, so I've been using Obsidian locally to do the work, I'd like to be doing on TiddlyWiki/Host.

**Comment by vettexl** [comment]
chiming in on this. tiddlywiki 5.3.1 using Edge 128.0.2739.54

**Comment by simonbaird** [comment]
So I can't reproduce this. In my experience saving takes a few seconds. Also I can't see any server-side errors that might be related to a save failure.

One theory is that this problem is specific to Windows users. Has any non-Windows user seen this problem? Are you able to try to reproduce it on a non-Windows computer, e.g. a Mac and see what happens. Does it happen across different networks (e.g. office, home, Starbucks), and across different computers?

Another question: Is it happening for any site or just certain sites? E.g. if you create a brand new site and click save, does it save quickly at that point? If you're trying this, could you also try with a Feather Wiki site to see if you're seeing save timeouts there as well.

Thanks very much for filing this and for your patience and persistence. I'd really like to figure out what's going on here.

**Comment by vettexl** [comment]
Started attempting to save around 21:30 PDT last night and finally got it to work around 22:05 PDT
I kept the window open and clicked on the red Save icon multiple times, eventually getting an error message once.

I will try again on my Windows machine and if not then I will attempt from a Linux VM

I know of at least three other people that have been having the same issue
maybe it has to do with the .tid files we're trying to upload?

i'm no developer so this is a bit out of my field

**Comment by vettexl** [comment]
hi Simon.

I tried a new TiddyWiki (5.3.5) site and even just changing the theme took several minutes (but eventually worked), nevermind uploading any tiddlers
tested the same using LibreWolf 125.0.3-1 on Windows, same save delay

I then tested with Firefox 130.0 on Ubuntu 22.04.4 LTS and saving worked immediately
tested the same using Edge 128.0.2739.67 on macOS Sonoma 14.5, saved immediately

something to do with Windows...

**Comment by simonbaird** [comment]
Thanks, that's really helpful.

**Comment by LordPalandus** [comment]
Just tried saving. 

Timed out after about 10 minutes and gave the error for timeout.

Tried with a brand new wiki. It saves fine the first time, within 5 seconds or so, but the second save without closing the browser, it times out again after roughly 10 minutes. And its only a single character changed, yet still times out. 

Tested out on Windows 8.1. Its possible the site no longer works for Windows 8.1. I had moved everything to Obsidian for local storage, as a temporary solution 2 weeks ago, but it looks like it will need to be a permanent solution until I get a new computer.

**Comment by simonbaird** [comment]
Here's one thing to try.

In the settings page for a site there's an option under "Advanced settings" called "Use the legacy upload saver for TiddlyWiki versions 5.2.3 and higher". Try checking that box, clicking "Update" to save the settings, and then try again with the saving and see if it makes any difference.

The goal is to see if the older "upload" saver works any better than the newer "put" saver. (See [here](https://github.com/simonbaird/tiddlyhost/wiki/About-the-two-different-save-methods) for info about the two different save methods.)

**Comment by vettexl** [comment]
the issue exists even without uploading anything as i mentioned earlier when i tried only changing theme settings
just now i tried only changing the title of the tiddlywiki and it just hangs
this was now with Arc browser 1.18.0.48077 (0000), Chromium Engine Version 128.0.6613.120 with Windows 10

this evening in Ubuntu i uploaded and saved files with the new "put" saver within seconds
that will have to be my workaround for now until you find the culprit

**Comment by simonbaird** [comment]
Did you try the "Use the legacy upload saver for TiddlyWiki versions 5.2.3 and higher" option @vettexl ?

**Comment by LordPalandus** [comment]
Could not find this legacy upload saver under Advanced or Settings.

I went to the control panel, hit advanced tab and checked under cascades, editor types, etc...

Then on the top bar I went to Settings, between Appearance and Saving, and couldn't find it there either.

There was also no setting in Saving called this either.

So... where am I supposed to look for it, under tiddlyhost?

**Comment by simonbaird** [comment]
Yeah, sorry I didn't explain it well. It's in the site settings for a site in Tiddlyhost, not in the TiddlyWiki itself. If you go to https://tiddlyhost.com/sites and click the first button in the "Actions" column it will take you there.

**Comment by LordPalandus** [comment]
So it saves instantly now with that change suggested. Tested it 5 times.

However, every time I hit save, I now download a file "tiddlywiki.html"

Now to delete 5 files.

**Comment by vettexl** [comment]
yes i tried that @simonbaird it did not seem to change anything for the better

**Comment by simonbaird** [comment]
> So it saves instantly (less than half a second) now with that change suggested. Tested it 5 times.
> 
> However, every time I hit save, I now download a file "tiddlywiki.html"
> 
> Now to delete 5 files.

Beware, usually if you're downloading a file on save, it means you're not saving to Tiddlyhost.

Are you using a version of TiddlyWiki 5.1.23 or older? If so then you need to do one more thing, [see the details here](https://github.com/simonbaird/tiddlyhost/wiki/Using-a-5.1.23-or-earlier-version-of-TiddlyWiki).

**Comment by LordPalandus** [comment]
So...

I checked this setting = Use the legacy "upload" saver for TiddlyWiki versions 5.2.3 and higher

and the site's version is = 5.3.3

And you were right, the change I made did not get saved. Just downloaded the file.

Guess it looks like Obsidian will be my permanent solution, until I can upgrade to latest windows.

**Comment by jalilahlee** [comment]
Hi 
I am not a technical user of Tiddlyhost but I have been using it for about 3-4 years. 
about 2 months ago, I have started to face the same problem trying to save my work and it has most times errored out and a few times saved a basic file after more  than 30minutes waiting. 

I am running Tiddlyhost on Windows 10 Pro version 22H2 tested a new TiddlyWiki (self-contained) 5.3.5 vanilla file.

I have selected  the "Use the legacy upload saver for TiddlyWiki versions 5.2.3 and higher" option and it still  does not work.

**Comment by fu-sen** [comment]
Hi, I was reporting this issue here:
https://github.com/simonbaird/tiddlyhost/issues/342#issuecomment-2297737704

Since then I've done a few things and learned:
- TiddlyWiki files with large file sizes may time out and not be able to be saved. Feather Wiki files with small file sizes will be saved successfully.
- If you update the Wiki offline and use Upload, it will take longer but the update will be more likely to be successful.

My main PC is also Windows 10, but this problem occurs regardless of the OS. This problem occurs regardless of which version of TiddlyWiki I'm using.

**Comment by Leilei332** [comment]
Now the saving speed has significantly improved, it takes only 3 seconds to save a 2MB wiki. Before that saving a 4.3MB wiki took about 16 minutes.

![图片](https://github.com/user-attachments/assets/5658d0ef-63cf-4f95-a47b-89efb0235032)

**Comment by simonbaird** [comment]
Some recent stats:

```
       OS     Avg  Median  >10s  >20s  >30s     Max  Count
       --     ---  ------  ----  ----  ----     ---  -----
      All    6.42    2.11   13%    6%    3%  351.99   2326
  Windows    6.00    2.08   11%    5%    3%  228.15    723
    Linux    6.81    2.82   11%    7%    5%  166.12    615
    Other    6.30    2.78   17%    5%    1%  268.54    612
      Mac    6.02    0.75    8%    2%    2%  351.99    215
     iPad    1.14    1.12    0%    0%    0%    2.09     78
   iPhone   17.23   11.94   55%   31%   18%  102.35     65
  Android    2.20    0.85    6%    6%    0%   23.81     16
     curl    6.56    6.64    0%    0%    0%    6.64      2

```

Conclusions:
* Clearly there are still some saves taking way too long
* With the median being around 2 seconds, it's maybe not as bad as it has been
* I don't think there's evidence for it being a Windows specific problem, at least not currently

This data is from the last couple of days. Unfortunately I don't have older log data, so I can't produce the stats over a longer period.

FWIW I'm still not able to reproduce this problem, and I still have no idea about the root cause, but I did upgrade the server to a newer version of its operating system, which might have somehow made a difference.

**Comment by LordPalandus** [comment]
Well, I personally moved on to local storage via Obsidian to keep track of my wikis, as I've not been able to get save times less than 1 minute in length, so I've not used Tiddly much the past few weeks. 

However, based on the data, why is Ipad so fast? 

Maybe there is specific internet protocols it has, that other OSes lack, and that is why its so fast. If that is the case, looking into porting something similar to the other options, might help.

Or maybe the other OSes are tripping a virus or malware scanner on the server, and thus it takes added time to assess the data stream to ensure it doesn't contain malware. 

Maybe sending out a message to all users to try doing saving at a specific day / time, and then observing the server for that time period to assess if something specific is causing people to have slow save times. Kind of like a server-test, that some gaming companies do to ensure that launch day servers are functional on day one.
